fish-rust/src/wutil/mod.rs

```diff
@@ -50,6 +496,324 @@ pub fn fish_iswalnum(c: char) -> bool {
+        DT_REG => Some(DirEntryType::reg),
+        DT_LNK => Some(DirEntryType::lnk),
+        DT_SOCK => Some(DirEntryType::sock),
+        // todo! whiteout
+        _ => None,
+    }
+}
+
+fn stat_mode_to_entry_type(m: libc::mode_t) -> Option<DirEntryType> {
+    match m & S_IFMT {
+        S_IFIFO => Some(DirEntryType::fifo),
+        S_IFCHR => Some(DirEntryType::chr),
+        S_IFDIR => Some(DirEntryType::dir),
+        S_IFBLK => Some(DirEntryType::blk),
+        S_IFREG => Some(DirEntryType::reg),
+        S_IFLNK => Some(DirEntryType::lnk),
+        S_IFSOCK => Some(DirEntryType::sock),
+        _ => {
+            // todo! whiteout
+            None
+        }
+    }
+}
+
+/// Class for iterating over a directory, wrapping readdir().
+/// This allows enumerating the contents of a directory, exposing the file type if the filesystem
+/// itself exposes that from readdir(). stat() is incurred only if necessary: if the entry is a
+/// symlink, or if the caller asks for the stat buffer.
+/// Symlinks are followed.
+pub struct DirIter {
+    /// Whether this dir_iter considers the "." and ".." filesystem entries.
+    withdot: bool,
+
+    dir: *mut libc::DIR,
+    error: libc::c_int,
+    entry: DirEntry,
+}
+
+impl DirIter {
+    /// Open a directory at a given path. On failure, \p error() will return the error code.
+    /// Note opendir is guaranteed to set close-on-exec by POSIX (hooray).
+    pub fn new(path: &wstr, withdot: bool) -> Self {
+        let mut error = 0;
+        let dir = wopendir(path);
+        if dir.is_null() {
+            error = errno::errno().0;
+        }
+        let entry = DirEntry {
+            dirfd: unsafe { libc::dirfd(dir) },
+            ..Default::default()
+        };
+        DirIter {
+            withdot,
+            dir,
+            error,
+            entry,
+        }
+    }
+
+    /// Rewind the directory to the beginning.
+    pub fn rewind(&mut self) {
+        if self.dir.is_null() {
+            unsafe { libc::rewinddir(self.dir) };
+        }
+    }
+
+    pub fn next(&mut self) -> Option<&DirEntry> {
+        if self.dir.is_null() {
+            return None;
+        }
+        errno::set_errno(errno::Errno(0));
+        let dent = unsafe { libc::readdir(self.dir) };
+        if dent.is_null() {
+            self.error = errno::errno().0;
+            return None;
+        }
+        let dent = unsafe { &*dent };
+        // Skip . and ..,
+        // unless we've been told not to.
+        if !self.withdot
+            && [
+                &[b'.' as i8, b'\0' as i8, b'\0' as i8][..],
+                &[b'.' as i8, b'.' as i8, b'\0' as i8][..],
+            ]
+            .contains(&&dent.d_name[..3])
+        {
+            return self.next();
+        }
+
+        self.entry.reset();
+        let d_name: Vec<u8> = dent.d_name.iter().map(|b| *b as u8).collect();
+        self.entry.name = cstr2wcstring(&d_name);
+        self.entry.inode = dent.d_ino;
```

Comment:
I'm confused as to why this fails on FreeBSD - is libc incomplete there?

Reply:
looks like it's called `d_ino` on some platforms (https://docs.rs/libc/latest/libc/struct.dirent.html)
but `d_fileno` [on BSD](https://rust-lang.github.io/libc/x86_64-unknown-freebsd/doc/libc/struct.dirent.html)
which is totally weird because `fileno()` is a libc function for something totally different

Reply:
wtfudge

Reply:
This goes back to https://github.com/rust-lang/libc/pull/24, which called it d_fileno on freebsdlike and openbsdlike. I don't see an explanation.

(NetBSD was added later and also received "d_fileno")

Reply:
`d_fileno` is also the name documented in the FreeBSD [man page for dirent](https://manpage.me/?q=dirent)
Not sure how we could use `d_ino` on all platforms in C++, maybe the BSD C library defines d_ino for compatibillity

Reply:
1. I hate posix.
2. Seems we need one of those `#cfg` thingamabobs here